README.md
# maestro remote

Ce programme permet de commander un poêle à granule, basé sur la technologie Maestro, en s'interfaçant directement avec les serveurs MCZ.

Une documentation plus complète est disponible dans le directory doc ou via es url

Doc: https://henribi.github.io/Maestro_Remote/

Changelog: https://henribi.github.io/Maestro_Remote/changelog


docs/changelog.md
# Changelog maestro remote

>**IMPORTANT**
>
>Pour rappel s'il n'y a pas d'information sur la mise à jour, c'est que celle-ci concerne uniquement de la mise à jour de documentation, de traduction ou de texte

....

# 04/11/2021
- dans maestro.service, modification du type vers simple
- suppression de appels thread (memory leak !!)
- déplacement des fichiers de log vers /var/log/maestro
- ajout dans la fonction "on_message_mqtt" du support pour le changement d'heure

# pre 30/10/2021
- basé sur les travaux de Anthony L., EtienneME, SkyPhil, Pipolas
- déplacement dans le fichier \_config\_.py file des parametres du poele



_data_.py
#coding: utf-8

'''
Tables des correspondances

	Le rang 0 correspond à la position de l'information dans la trame MAESTRO
	Le rang 1 correspond a l'intitulé publié sur le broker
	Le rang 2 (optionnel) permet de remplacer le code de la trame par une information texte correspondante

'''
RecuperoInfo=[
	[1,"Etat du poele",[
						[0, "Eteint"],
						[1, "Controle du poele froid / chaud"],
						[2, "Clean Froid"],
						[3, "Load Froid"],
						[4, "Start 1 Froid"],
						[5, "Start 2 Froid"],
						[6, "Clean Chaud"],
						[7, "Load Chaud"],
						[8, "Start 1 chaud"],
						[9, "Start 2 chaud"],
						[10, "Stabilisation"],
						[11, "Puissance 1"],
						[12, "Puissance 2"],
						[13, "Puissance 3"],
						[14, "Puissance 4"],
						[15, "Puissance 5"],
						[30, "Mode diagnostique"],
						[31, "Marche"],
						[40, "Extinction"],
						[41, "Refroidissement en cours"],
						[42, "Nettoyage basse p."],
						[43, "Nettoyage haute p."],
						[44, "Debloquage vis sans fin"],
						[45, "AUTO ECO"],
						[46, "Standby"],
						[48, "Diagnostique"],
						[49, "CHARG. VIS SANS FIN"],
						[50, "Erreur A01 - Allumage rate"],
						[51, "Erreur A02 - Pas de flamme"],
						[52, "Erreur A03 - Surchauffe du reservoir"],
						[53, "Erreur A04 - Temperature des fumees trop haute"],
						[54, "Erreur A05 - Obstruction conduit - Vent"],
						[55, "Erreur A06 - Mauvais tirage"],
						[56, "Erreur A09 - Defaillance sonde de fumees"],
						[57, "Erreur A11 - Defaillance motoreducteur"],
						[58, "Erreur A13 - Temperature carte mere trop haute"],
						[59, "Erreur A14 - Defaut Active"],
						[60, "Erreur A18 - Temperature d'eau trop haute"],
						[61, "Erreur A19 - Defaut sonde temperature eau"],
						[62, "Erreur A20 - Defaut sonde auxiliaire"],
						[63, "Erreur A21 - Alarme pressostat"],
						[64, "Erreur A22 - Defaut sonde ambiante"],
						[65, "Erreur A23 - Defaut fermeture brasero"],
						[66, "Erreur A12 - Panne controleur motoreducteur"],
						[67, "Erreur A17 - Bourrage vis sans fin"],
						[69, "Attente Alarmes securite"],
						]],
	[2,"Etat du ventilateur d'ambiance",[
										[0, "Desactive"],
										[1, "Niveau 1"],
										[2, "Niveau 2"],
										[3, "Niveau 3"],
										[4, "Niveau 4"],
										[5, "Niveau 5"],
										[6, "Automatique"],
										]],
	[3,"Etat du ventilateur canalise 1",[
										[0, "Desactive"],
										[1, "Niveau 1"],
										[2, "Niveau 2"],
										[3, "Niveau 3"],
										[4, "Niveau 4"],
										[5, "Niveau 5"],
										[6, "Automatique"],
										]],
	[4,"Etat du ventilateur canalise 2",[
										[0, "Désactive"],
										[1, "Niveau 1"],
										[2, "Niveau 2"],
										[3, "Niveau 3"],
										[4, "Niveau 4"],
										[5, "Niveau 5"],
										[6, "Automatique"],
										]],
	[5,"Temperature des fumees"],
	[6,"Temperature ambiante"],
	[7,"Puffer Temperature"], # !=255 == Hydro
	[8,"Temperature chaudiere"],
	[9,"Temperature NTC3"], # !=255 == Hydro
	[10,"Etat de la bougie",[
					[0, "Ok"],
					[1, "Usee"],
					]],
	[11,"ACTIVE - Set"],
	[12,"RPM - Ventilateur fummees"],
	[13,"RPM - Vis sans fin - SET"],
	[14,"RPM - Vis sans fin - LIVE"],
	[17,"Brazero",[
					[0, "OK"],
					[101, "Ouverture brazero"],
					[100, "Fermeture brazero"],
					]], # !==Matic
	[18,"Profil",[
					[0, "Manuel"],
					[1, "Dynamic"],
					[2, "Overnight"],
					[3, "Comfort"],
					[4, "Power 110%"],
					[10, "Mode Adaptatif"],
					]],
	[20,"Etat du mode Active",[
					[0, "Off"],
					[1, "On"],
					]],  #0: Désactivé, 1: Activé
	[21,"ACTIVE - Live"],
	[22,"Mode de regulation",[
							[0, "Manuelle"],
							[1, "Dynamique"],
							]],
	[23,"Mode ECO",[
					[0, "Off"],
					[1, "On"],
					]],
	[24,"Silence",[
					[0, "Off"],
					[1, "On"],
					]],
	[25,"Mode Chronotermostato",[
					[0, "Off"],
					[1, "On"],
					]],
	[26,"TEMP - Consigne"],
	[27,"TEMP - Boiler"],
	[28,"TEMP - Carte mere"],
	[29,"Puissance Active",[
							[11, "Puissance 1"],
							[12, "Puissance 2"],
							[13, "Puissance 3"],
							[14, "Puissance 4"],
							[15, "Puissance 5"],
							]],
	[32,"Heure du poele (0-23)"],
	[33,"Minutes du poele (0-29)"],
	[34,"Jour du poele (1-31)"],
	[35,"Mois du poele (1-12)"],
	[36,"Année du poele"],
	[37,"Heures de fonctionnement total (s)"],
	[38,"Heures de fonctionnement en puissance 1 (s)"],
	[39,"Heures de fonctionnement en puissance 2 (s)"],
	[40,"Heures de fonctionnement en puissance 3 (s)"],
	[41,"Heures de fonctionnement en puissance 4 (s)"],
	[42,"Heures de fonctionnement en puissance 5 (s)"],
	[43,"Heures avant entretien"],
	[44,"Minutes avant extinction"],
	[45,"Nombre d'allumages"],
	[47,"Sonde Pellets",[
						[0, "Sonde pas active"],
						[10, "Niveau suffisant"],
						[11, "Niveau presque vide"],
						]],
	[48,"Effet sonore",[
					[0, "Off"],
					[1, "On"],
					]],
	[49,"Etat effets sonores",[
					[0, "Off"],
					[1, "On"],
					]],
	[50,"Sleep",[
					[0, "Off"],
					[1, "On"],
					]],
	[51,"Mode",[
				[0, "Hiver"],
				[1, "Ete"],
				]],
	[52,"Sonde wifi temperature 1"],
	[53,"Sonde wifi temperature 2"],
	[54,"Sonde wifi temperature 3"],
	[55,"Inconnu"],
	[56,"Set Puffer"],
	[57,"Set Boiler"],
	[58,"Set Health"], # !==Hydro
	[59,"Temperature retour"],
	[60,"Antigel",[
					[0, "Off"],
					[1, "On"],
					]],
	]

maestro.py
#!/usr/bin/python3
# coding: utf-8

# Basé sur les travaux de Anthony L., EtienneME, Pipolas
#
# Modified version
# déplacement dans config file des parametres du poele
# suppression de appels thread (memory leak !!)
# ajout dans fonction "on_message_mqtt" du support de changement d'heure
#

import paho.mqtt.client as mqtt
import time
import sys
import os
import json
import logging
import datetime
from logging.handlers import RotatingFileHandler, TimedRotatingFileHandler
import websocket

from pprint import pprint
import socketio

from _config_ import  _MCZ_device_serial
from _config_ import  _MCZ_device_MAC
from _config_ import  _MCZ_App_URL

# MQTT
from _config_ import _MQTT_ip
from _config_ import _MQTT_port
from _config_ import _MQTT_TOPIC_SUB
from _config_ import _MQTT_TOPIC_PUB

from _config_ import _MQTT_authentication
from _config_ import _MQTT_user
from _config_ import _MQTT_pass


try:
    import thread
except ImportError:
    import _thread as thread

#fpid =os.fork()
#if fpid!=0:
#    # Running as daemon now. PID is fpid
#    sys.exit(0)

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s :: %(levelname)s :: %(message)s')
file_handler = TimedRotatingFileHandler('/var/log/maestro/maestro.log', when='D', interval=1, backupCount=5)
file_handler.setLevel(logging.DEBUG)
file_handler.setFormatter(formatter)
logger.addHandler(file_handler)

stream_handler = logging.StreamHandler()
stream_handler.setLevel(logging.DEBUG)
logger.addHandler(stream_handler)


class PileFifo(object):
    def __init__(self, maxpile=None):
        self.pile = []
        self.maxpile = maxpile

    def empile(self, element, idx=0):
        if (self.maxpile != None) and (len(self.pile) == self.maxpile):
            raise ValueError("erreur: tentative d'empiler dans une pile pleine")
        self.pile.insert(idx, element)

    def depile(self, idx=-1):
        if len(self.pile) == 0:
            raise ValueError("erreur: tentative de depiler une pile vide")
        if idx < -len(self.pile) or idx >= len(self.pile):
            raise ValueError("erreur: element de pile à depiler n'existe pas")
        return self.pile.pop(idx)

    def element(self, idx=-1):
        if idx < -len(self.pile) or idx >= len(self.pile):
            raise ValueError("erreur: element de pile à lire n'existe pas")
        return self.pile[idx]

    def copiepile(self, imin=0, imax=None):
        if imax == None:
            imax = len(self.pile)
        if imin < 0 or imax > len(self.pile) or imin >= imax:
            raise ValueError("erreur: mauvais indice(s) pour l'extraction par copiepile")
        return list(self.pile[imin:imax])

    def pilevide(self):
        return len(self.pile) == 0

    def pilepleine(self):
        return self.maxpile != None and len(self.pile) == self.maxpile

    def taille(self):
        return len(self.pile)


Message_MQTT = PileFifo()
Message_WS = PileFifo()

# SIO CONNECT TO MCZ MAESTRO
sio = socketio.Client(logger=True, engineio_logger=True)


_INTERVALLE = 1
_TEMPS_SESSION = 60


MQTT_MAESTRO = {}

logger.info('Lancement du deamon')
logger.info('Anthony L. 2019; Henribi 2021')
logger.info('Niveau de LOG : DEBUG')


def send():
    #def run(*args):
        time.sleep(_INTERVALLE)
        if Message_MQTT.pilevide():
            Message_MQTT.empile("C|RecuperoInfo")
        cmd = Message_MQTT.depile()
        logger.info("Envoi de la commande : " + str(cmd))
        sio.emit(
            "chiedo",
            {
                "serialNumber": _MCZ_device_serial,
                "macAddress": _MCZ_device_MAC,
                "tipoChiamata": 1,
                "richiesta": cmd,
            },
        )

    #run()


def on_connect_mqtt(client, userdata, flags, rc):
    logger.info("Connecté au broker MQTT avec le code : " + str(rc))


def on_message_mqtt(client, userdata, message):
    logger.info('Message MQTT reçu : ' + str(message.payload.decode()))
    cmd = message.payload.decode().split(",")
    if (int(cmd[0])) < 9000:
        if cmd[0] == "42":
            cmd[1] = (int(cmd[1]))
        Message_MQTT.empile("C|WriteParametri|" + cmd[0] + "|" + str(cmd[1]))
        logger.info('Contenu Pile Message_MQTT : ' + str(Message_MQTT.copiepile()))
        send()
    else:
        if cmd[0] == "9001":
            order = "C|SalvaDataOra|"
        Message_MQTT.empile(str(order) + str(cmd[1]))
        logger.info('Contenu Pile Message_MQTT : ' + str(Message_MQTT.copiepile()))
        send()



def secTOdhms(nb_sec):
    qm, s = divmod(nb_sec, 60)
    qh, m = divmod(qm, 60)
    d, h = divmod(qh, 24)
    return "%d:%d:%d:%d" % (d, h, m, s)


@sio.event
def connect():
    pprint("Connected")
    pprint("SID is : {}".format(sio.sid))
    sio.emit(
        "join",
        {
            "serialNumber": _MCZ_device_serial,
            "macAddress": _MCZ_device_MAC,
            "type": "Android-App",
        },
    )
    sio.emit(
        "chiedo",
        {
            "serialNumber": _MCZ_device_serial,
            "macAddress": _MCZ_device_MAC,
            "tipoChiamata": 0,
            "richiesta": "RecuperoParametri",
        },
    )
    sio.emit(
        "chiedo",
        {
            "serialNumber": _MCZ_device_serial,
            "macAddress": _MCZ_device_MAC,
            "tipoChiamata": 1,
            "richiesta": "C|RecuperoInfo",
        },
    )


@sio.event
def disconnect():
    pprint("Disconnected")


@sio.event
def rispondo(response):
    pprint("Received 'rispondo' message")
    datas = response["stringaRicevuta"].split("|")
    from _data_ import RecuperoInfo
    for i in range(0, len(datas)):
        for j in range(0, len(RecuperoInfo)):
            if i == RecuperoInfo[j][0]:
                if len(RecuperoInfo[j]) > 2:
                    for k in range(0, len(RecuperoInfo[j][2])):
                        if int(datas[i], 16) == RecuperoInfo[j][2][k][0]:
                            MQTT_MAESTRO[RecuperoInfo[j][1]] = RecuperoInfo[j][2][k][1]
                            break
                        else:
                            MQTT_MAESTRO[RecuperoInfo[j][1]] = ('Code inconnu :', str(int(datas[i], 16)))
                else:
                    if i == 5 or i == 6 or i == 7 or i == 8 or i == 9 or i == 26 or i == 27 or i == 28 or i == 46 or i == 52 or i == 53 or i == 54 or i == 59:
                    ###if i == 6 or i == 26 or i == 28:
                        MQTT_MAESTRO[RecuperoInfo[j][1]] = float(int(datas[i], 16)) / 2

                    elif i >= 37 and i <= 42:
                        MQTT_MAESTRO[RecuperoInfo[j][1]] = secTOdhms(int(datas[i], 16))
                    else:
                        MQTT_MAESTRO[RecuperoInfo[j][1]] = int(datas[i], 16)
    logger.info('Publication sur le topic MQTT ' + str(_MQTT_TOPIC_PUB) + ' le message suivant : ' + str(
        json.dumps(MQTT_MAESTRO)))
    client.publish(_MQTT_TOPIC_PUB, json.dumps(MQTT_MAESTRO), 1)


sio.connect(_MCZ_App_URL)

logger.info('Connection en cours au broker MQTT (IP:' + _MQTT_ip + ' PORT:' + str(_MQTT_port) + ')')
client = mqtt.Client()
if _MQTT_authentication == True:
    client.username_pw_set(username=_MQTT_user, password=_MQTT_pass)
client.on_connect = on_connect_mqtt
client.on_message = on_message_mqtt
client.connect(_MQTT_ip, _MQTT_port)
client.loop_start()
logger.info('Souscription au topic ' + str(_MQTT_TOPIC_SUB) + ' avec un Qos=1')
client.subscribe(_MQTT_TOPIC_SUB, qos=1)


def receive(*args):
    while True:
        time.sleep(30)
        logger.info("Envoi de la commande pour rafraichir les donnees")
        sio.emit(
            "chiedo",
            {
                "serialNumber": _MCZ_device_serial,
                "macAddress": _MCZ_device_MAC,
                "tipoChiamata": 1,
                "richiesta": "C|RecuperoInfo",
            },
        )
    time.sleep(15)


receive()

